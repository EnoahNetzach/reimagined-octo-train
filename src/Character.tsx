import { useTranslation } from 'react-i18next'
import { useParams } from 'react-router-dom'
import { useQuery } from '@apollo/client'
import { gql } from './__generated__'
import getCharacterStatusColor from './getCharacterStatusColor.tsx'

const GET_CHARACTER = gql`
  query GetCharacter($id: ID!) {
    character(id: $id) {
      id
      name
      status
      gender
      species
      type
      image
      origin {
        id
        name
        type
        dimension
        residents {
          id
        }
      }
      location {
        id
        name
        type
        dimension
        residents {
          id
        }
      }
      episode {
        id
        name
        air_date
      }
    }
  }
`

export default function Character() {
  const { t } = useTranslation()
  const { characterId } = useParams()
  const { data, loading } = useQuery(GET_CHARACTER, { variables: { id: characterId } })
  const { character } = data ?? {}

  return (
    <article className="isolate grid gap-2 grid-cols-1">
      {loading ? (
        t('loading')
      ) : (
        <>
          <img alt="" className="object-cover max-h-96 max-w-96" src={character.image} />

          <h1 className="text-6xl font-bold text-primary">{character.name}</h1>

          <section>
            <span
              aria-description={t('character.status')}
              className={`inline-block rounded-full w-2 h-2 ${getCharacterStatusColor(character)}`}
            />{' '}
            {t(`character.status.${character.status?.toLowerCase()}`)}
          </section>

          <section>
            <h3 className="text-secondary inline-block">{t('character.species')}</h3> {character.species}
            {character.type ? <span> ({character.type})</span> : null}
          </section>

          <section>
            <h3 className="text-secondary inline-block">{t('character.gender')}</h3>{' '}
            {t(`character.gender.${character.gender?.toLowerCase()}`)}
          </section>

          <section>
            <h3 className={`text-secondary ${character.origin.id ? '' : 'inline-block'}`}>{t('origin.name')}</h3>

            {character.origin.id ? (
              <details>
                <summary>{character.origin.name}</summary>

                <div className="py-0 px-4">
                  <h4 className="text-ternary inline-block">{t('location.type')}</h4> {character.origin.type}
                  <br />
                  <h4 className="text-ternary inline-block">{t('location.dimension')}</h4>{' '}
                  {character.origin.dimension.toLowerCase() === 'unknown'
                    ? t('dimension.unknown')
                    : character.origin.dimension}
                  <br />
                  <h4 className="text-ternary inline-block">{t('location.residents')}</h4>{' '}
                  {character.origin.residents.length}
                </div>
              </details>
            ) : (
              ' ' + t('origin.unknown')
            )}
          </section>

          <section>
            <h3 className={`text-secondary ${character.location.id ? '' : 'inline-block'}`}>{t('location.name')}</h3>

            {character.location.id ? (
              <details>
                <summary>{character.location.name}</summary>

                <div className="py-0 px-4">
                  <h4 className="text-ternary inline-block">{t('location.type')}</h4> {character.location.type}
                  <br />
                  <h4 className="text-ternary inline-block">{t('location.dimension')}</h4>{' '}
                  {character.location.dimension.toLowerCase() === 'unknown'
                    ? t('dimension.unknown')
                    : character.location.dimension}
                  <br />
                  <h4 className="text-ternary inline-block">{t('location.residents')}</h4>{' '}
                  {character.location.residents.length}
                </div>
              </details>
            ) : (
              ' ' + t('location.unknown')
            )}
          </section>

          <section>
            <h3 className="text-secondary">{t('episode.list')}</h3>
            <details>
              <summary>{t('episode.number', { count: character.episode.length })}</summary>

              {character.episode.map((episode) => (
                <div className="py-0 px-4" key={episode.id}>
                  <h4 className="text-ternary inline-block">{episode.name}</h4>{' '}
                  {t('episode.aired', {
                    formatParams: { val: { year: 'numeric', month: 'short', day: 'numeric' } },
                    val: new Date(Date.parse(episode.air_date)),
                  })}
                </div>
              ))}
            </details>
          </section>
        </>
      )}
    </article>
  )
}
