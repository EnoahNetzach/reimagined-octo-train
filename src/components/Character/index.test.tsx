import { createRenderer } from 'react-test-renderer/shallow'
import * as apolloClient from '@apollo/client'
import { QueryResult } from '@apollo/client'
import { render } from '@testing-library/react'
import Character from '.'

const mockCharacter = {
  episode: [
    {
      air_date: '__character_episode_air_date__',
      id: '__character_episode_id__',
      name: '__character_episode_name__',
    },
  ],
  gender: '__character_gender__',
  id: '__character_id__',
  image: '__character_image__',
  location: {
    dimension: '__character_location_dimension__',
    id: '__character_location_id__',
    name: '__character_location_name__',
    residents: Array.from({ length: 5 }),
    type: '__character_location_type__',
  },
  name: '__character_name__',
  origin: {
    dimension: '__character_origin_dimension__',
    id: '__character_origin_id__',
    name: '__character_origin_name__',
    residents: Array.from({ length: 6 }),
    type: '__character_origin_type__',
  },
  species: '__character_species__',
  status: '__character_status__',
  type: '__character_type__',
}

jest.mock('@apollo/client', () => ({
  useQuery: jest.fn(() => ({
    data: {
      character: mockCharacter,
    },
    loading: false,
  })),
}))

jest.mock('~/__generated__', () => ({ gql: jest.fn() }))

jest.mock('~/utils/getCharacterStatusColor', () => () => '__character_status_color__')

describe('Character', () => {
  it('mounts', () => {
    const renderer = createRenderer()
    renderer.render(<Character />)
    expect(renderer.getRenderOutput()).toMatchSnapshot()
  })

  it('loads', () => {
    jest.spyOn(apolloClient, 'useQuery').mockReturnValueOnce({
      loading: true,
    } as QueryResult<unknown>)

    const renderer = createRenderer()
    renderer.render(<Character />)
    expect(renderer.getRenderOutput()).toMatchSnapshot()
  })

  it('has the correct data', () => {
    const { queryByText } = render(<Character />)
    expect(queryByText('__character_name__')).toBeInTheDocument()
    expect(queryByText('__character_species__')).toBeInTheDocument()
    expect(document.querySelector('img[src="__character_image__"]')).not.toBeNull()
    expect(queryByText('__character_origin_name__')).toBeInTheDocument()
    expect(queryByText('t(origin.unknown)')).not.toBeInTheDocument()
    expect(queryByText('__character_location_name__')).toBeInTheDocument()
    expect(queryByText('t(location.unknown)')).not.toBeInTheDocument()
  })

  it('has the correct data when the location and origin are unknown', () => {
    jest.spyOn(apolloClient, 'useQuery').mockReturnValueOnce({
      data: {
        character: {
          ...mockCharacter,
          origin: null,
          location: null,
        },
      },
      loading: false,
    } as QueryResult<unknown>)

    const { queryByText } = render(<Character />)
    expect(queryByText('__character_origin_name__')).not.toBeInTheDocument()
    expect(queryByText('t(origin.unknown)')).toBeInTheDocument()
    expect(queryByText('__character_location_name__')).not.toBeInTheDocument()
    expect(queryByText('t(location.unknown)')).toBeInTheDocument()
  })
})
