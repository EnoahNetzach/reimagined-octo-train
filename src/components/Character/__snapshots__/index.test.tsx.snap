// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Character loads 1`] = `
<article
  className="isolate grid gap-2 grid-cols-1"
>
  <div
    className="p-8 flex justify-center"
  >
    <Loading />
  </div>
</article>
`;

exports[`Character mounts 1`] = `
<article
  className="isolate grid gap-2 grid-cols-1"
>
  <React.Fragment>
    <img
      alt=""
      className="object-cover h-48 w-48 lg:h-96 lg:w-96"
      fetchpriority="high"
      src="__character_image__"
    />
    <div
      className="lg:contents"
    >
      <h1
        className="heading-title-bold-primary"
      >
        __character_name__
      </h1>
      <section>
        <span
          aria-description="t(character.status)"
          className="inline-block rounded-full w-2 h-2 __character_status_color__"
        />
         
        t(character.status.__character_status__)
      </section>
      <section>
        <h3
          className="heading-title-tertiary inline-block"
        >
          t(character.species)
        </h3>
         
        __character_species__
        <span>
           (
          __character_type__
          )
        </span>
      </section>
      <section>
        <h3
          className="heading-title-tertiary inline-block"
        >
          t(character.gender)
        </h3>
         
        t(character.gender.__character_gender__)
      </section>
      <section>
        <h3
          className="heading-title-tertiary "
        >
          t(origin.name)
        </h3>
        <details>
          <summary>
            __character_origin_name__
          </summary>
          <div
            className="py-0 px-4"
          >
            <h4
              className="text-ternary inline-block"
            >
              t(location.type)
            </h4>
             
            __character_origin_type__
            <br />
            <h4
              className="text-ternary inline-block"
            >
              t(location.dimension)
            </h4>
             
            __character_origin_dimension__
            <br />
            <h4
              className="text-ternary inline-block"
            >
              t(location.residents)
            </h4>
             
            6
          </div>
        </details>
      </section>
      <section>
        <h3
          className="heading-title-tertiary "
        >
          t(location.name)
        </h3>
        <details>
          <summary>
            __character_location_name__
          </summary>
          <div
            className="py-0 px-4"
          >
            <h4
              className="text-ternary inline-block"
            >
              t(location.type)
            </h4>
             
            __character_location_type__
            <br />
            <h4
              className="text-ternary inline-block"
            >
              t(location.dimension)
            </h4>
             
            __character_location_dimension__
            <br />
            <h4
              className="text-ternary inline-block"
            >
              t(location.residents)
            </h4>
             
            5
          </div>
        </details>
      </section>
      <section>
        <h3
          className="heading-title-tertiary"
        >
          t(episode.list)
        </h3>
        <details>
          <summary>
            t(episode.number)
          </summary>
          <div
            className="py-0 px-4"
          >
            <h4
              className="text-ternary inline-block"
            >
              __character_episode_name__
            </h4>
             
            t(episode.aired)
          </div>
        </details>
      </section>
    </div>
  </React.Fragment>
</article>
`;
